src/coreclr/jit/lower.cpp

```diff
@@ -2936,16 +2936,24 @@ GenTree* Lowering::OptimizeConstCompare(GenTree* cmp)
                 if (castOp->OperIs(GT_OR, GT_XOR, GT_AND))
                 {
                     GenTree* op1 = castOp->gtGetOp1();
-                    if ((op1 != nullptr) && !op1->IsCnsIntOrI())
+                    GenTree* op2 = castOp->gtGetOp2();
+
+#ifdef TARGET_XARCH
+                    // If one of the ops is a GT_LCL_VAR, then change its type to the castToType.
+                    // We do this to take advantage of containment for a memory op.
+                    if (op1->OperIs(GT_LCL_VAR) && !op2->IsCnsIntOrI())
                     {
-                        op1->ClearContained();
+                        op1->ChangeType(castToType);
```

Comment:
It is not legal to retype "any" local like that. I think it risks at least confusing spilling code.

Reply:
A little unfortunate, as making this contained would turn something like this:
```
mov reg0, ptr
and reg1, reg0
``` 
to
```
and reg1, ptr
```

I would need to think through an alternative way that doesn't need to change the type. Maybe relaxing the containment to accept larger values - though I think I tried this originally and there were other problems.

In this case, the `AND`, `OR`, and `XOR` ops get retyped themselves to the small type, I believe this is to take advantage of the CPU flags getting set from those operations about the result so the `JCC` ops can act accordingly, otherwise it would have to do a `cmp`. This already happens today, so we would end up having a node that looks like this:

`AND{byte}(VAR{int}, VAR{int})` 

Is that tree above legal?

Reply:
> Is that tree above legal?

Not in general, no (but it is "fine" in this very particular case). This transformation is just a little hacky in how it does things.

Reply:
Right, it does seem to be fine. By now you can understand why these rules and edge cases might all be confusing to me :)